api/users.rb

```diff
@@ -11,6 +11,50 @@
   end
 end
 
+get "#{APIPREFIX}/users/:course_id/stats2" do |course_id|
+  page = (params["page"] || DEFAULT_PAGE).to_i
+  page = [1, page].max
+  per_page = (params["per_page"] || DEFAULT_PER_PAGE).to_i
+  per_page = DEFAULT_PER_PAGE if per_page <= 0
+
+  # There are two sorts available, activity t sor
+  sort_by = params["sort_key"]
+  if sort_by == "flagged"
+    # If sorting by flags we sort by active flags and then inactive flags
+    sort_criterion = [
+      ["course_stats.active_flags", :desc],
+      ["course_stats.inactive_flags", :desc],
+    ]
+  else
+    # If sorting by activity (default) sort by thread count, then responses, then replies.
+    sort_criterion = [
+      ["course_stats.threads", :desc],
+      ["course_stats.responses", :desc],
+      ["course_stats.replies", :desc],
+    ]
+  end
+
+  paginated_stats = User
+                      .where("course_stats.course_id" => course_id)
+                      .only(:username, :'course_stats.$')
+                      .order_by(sort_criterion)
+                      .paginate(:page => page, :per_page => per_page)
+  total_count = paginated_stats.total_entries
+
+  data = paginated_stats.to_a.map { |user_stats| {
+    :username => user_stats["username"]
+  }.merge(
+    user_stats["course_stats"].first.except("_id", "course_id")
+  ) }
```

Comment:
Flatten the nesting, and remove the ids from each document since the id is unneeded, and the course_id is common.